tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -226,6 +226,296 @@ static bool HasCanonicalDimensionNumbers(
+    // einsum_config = "lhs_loop,rhs_loop->out_loop"
+    std::size_t pos_arrow = einsum_config.find(kArrow_);
+    std::size_t pos_comma = einsum_config.find(kComma_);
+
+    StringRef lhs_loop = einsum_config.substr(0, pos_comma);
+    StringRef rhs_loop = einsum_config.substr(
+        pos_comma + kComma_.size(), pos_arrow - (pos_comma + kComma_.size()));
+    StringRef out_loop = einsum_config.substr(pos_arrow + kArrow_.size());
+
+    // Check for Invalid Configs.
+    // 1.Check that there is only maximum 2 inputs
+    // 2.Check that there is only maximum 1 output
+    // 3.Check that there is 1 kArrow_
+    if (rhs_loop.find(kComma_) != std::string::npos ||
+        out_loop.find(kComma_) != std::string::npos ||
+        out_loop.find(kArrow_) != std::string::npos) {
+      return rewriter.notifyMatchFailure(op, "Invalid einsum config!");
+    }
+
+    // Find result type, if on tensors.
+    auto result_ty = this->typeConverter
+                         ->convertType(GetHloOpResultType</*isLHLO=*/false>(op))
+                         .dyn_cast<RankedTensorType>();
+
+    // Check result type compatibility.
+    if (!result_ty || !(result_ty.getElementType().isSignlessIntOrFloat())) {
+      return rewriter.notifyMatchFailure(op, "Invalid result type");
+    }
+
+    // Convert the representation to vector<string>.
+    SmallVector<std::string> lhs_ein =
+        GetEinsumConfigAsVector(lhs_loop, get_rank(adaptor.lhs()));
+    SmallVector<std::string> rhs_ein =
+        GetEinsumConfigAsVector(rhs_loop, get_rank(adaptor.rhs()));
+    SmallVector<std::string> out_ein =
+        GetEinsumConfigAsVector(out_loop, result_ty.getRank());
+    SmallVector<SmallVector<std::string>> all_loop_dim = {lhs_ein, rhs_ein,
+                                                          out_ein};
+
+    if (!CheckBatchHasEqualRank(lhs_ein.size(), lhs_loop, rhs_ein.size(),
+                                rhs_loop, out_ein.size(), out_loop)) {
+      return rewriter.notifyMatchFailure(
+          op, "Invalid elipsis('...') within einsum config!");
+    }
+
+    // Find all unique indices in the input and output.
+    llvm::SmallSetVector<StringRef, 4> input_ind;
+    llvm::SmallSetVector<StringRef, 4> output_ind;
+    for (const auto& loops_it : llvm::enumerate(all_loop_dim)) {
+      for (const StringRef& dim : loops_it.value()) {
+        if (loops_it.index() < args.size()) {
+          input_ind.insert(dim);
+        } else {
+          output_ind.insert(dim);
+        }
+      }
+    }
+
+    llvm::SmallSetVector<StringRef, 4> reduction_axe =
+        FindSummationAxes(input_ind, output_ind);
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+
+    // Prepare init tensor for linalg.generic op.
+    auto dyn_sizes = ExtractDynamicEinsumSizes(
+        rewriter, loc, adaptor.lhs(), adaptor.rhs(), lhs_ein, rhs_ein, out_ein);
+    Value output = GetInitTensor(rewriter, loc, result_ty, dyn_sizes);
+    if (reduction_axe.size() > 0) {
```

Comment:
should use `!reduction_axe.empty()` instead

Reply:
done